### ISSUE TITLE
Specifying only n-nodes without n-tasks causes wreck/pmi failures

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#411]
Having gotten some of the kinks worked out of profiling a broker, I went to run some MPI tests with it, and seem not to be able to run anything MPI-based on more than one node.

On a single node, mpi_hello works fine, though 3 seconds seems... unfortuante... for just four processes:
```
flux wreckrun -n 4 ./mpi_hello
lt-flux-broker: [1442534481.147908] job.info[0] got request job.create
lt-flux-broker: [1442534481.149268] job.info[0] Setting job 6 to reserved
lt-flux-broker: [1442534481.191392] lwj.6.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442534481.192444] lwj.6.debug[0] lwj.6: node0: basis=0
lt-flux-broker: [1442534481.196138] lwj.6.info[0] lwj 6: node0: nprocs=4, nnodes=1, cmdline=[ ".\/mpi_hello" ]
lt-flux-broker: [1442534481.196205] lwj.6.debug[0] updating job state to starting
lt-flux-broker: [1442534481.197697] lwj.6.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534481.202228] lwj.6.debug[0] task0: pid 242528 (./mpi_hello): started
lt-flux-broker: [1442534481.202427] lwj.6.debug[0] task1: pid 242529 (./mpi_hello): started
lt-flux-broker: [1442534481.202634] lwj.6.debug[0] task2: pid 242530 (./mpi_hello): started
lt-flux-broker: [1442534481.202831] lwj.6.debug[0] task3: pid 242531 (./mpi_hello): started
lt-flux-broker: [1442534481.204447] lwj.6.debug[0] updating job state to running
0: completed MPI_Init in 2.855s.  There are 4 tasks
lt-flux-broker: [1442534484.215292] lwj.6.debug[0] task0: pid 242528 (./mpi_hello) exited with status 0x0000
lt-flux-broker: [1442534484.215803] lwj.6.debug[0] task1: pid 242529 (./mpi_hello) exited with status 0x0000
lt-flux-broker: [1442534484.216323] lwj.6.debug[0] task2: pid 242530 (./mpi_hello) exited with status 0x0000
lt-flux-broker: [1442534484.216742] lwj.6.debug[0] task3: pid 242531 (./mpi_hello) exited with status 0x0000
lt-flux-broker: [1442534484.218629] lwj.6.debug[0] updating job state to complete
lt-flux-broker: [1442534484.220869] lwj.6.info[0] job complete. exiting...
0: completed first barrier
0: completed MPI_Finalize
lt-flux-broker: [1442534484.223804] job.info[0] got request job.disconnect
```

Running it on four nodes, one process per node, initially gave me one task on one node and PMGR collective errors on the other three, like this:
```
flux wreckrun -N 4 ./mpi_hello
lt-flux-broker: [1442534264.775650] job.info[0] got request job.create
lt-flux-broker: [1442534264.776906] job.info[0] Setting job 4 to reserved
lt-flux-broker: waitpid: No child processes
lt-flux-broker: [1442534264.810765] lwj.4.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442534264.811804] lwj.4.debug[0] lwj.4: node0: basis=0
lt-flux-broker: [1442534264.813179] lwj.4.info[0] lwj 4: node0: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello" ]
lt-flux-broker: [1442534264.813249] lwj.4.debug[0] updating job state to starting
lt-flux-broker: [1442534264.814502] lwj.4.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534264.818279] lwj.4.debug[0] task0: pid 242448 (./mpi_hello): started
lt-flux-broker: [1442534264.832277] lwj.4.debug[1] initializing from CMB: rank=1
lt-flux-broker: [1442534264.832419] lwj.4.debug[2] initializing from CMB: rank=2
lt-flux-broker: [1442534264.832824] lwj.4.debug[1] lwj.4: node1: basis=1
lt-flux-broker: [1442534264.832659] lwj.4.debug[3] initializing from CMB: rank=3
lt-flux-broker: [1442534264.833404] lwj.4.info[1] lwj 4: node1: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello" ]
lt-flux-broker: [1442534264.833608] lwj.4.debug[1] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534264.833839] lwj.4.debug[2] lwj.4: node2: basis=2
lt-flux-broker: [1442534264.834154] lwj.4.debug[3] lwj.4: node3: basis=3
lt-flux-broker: [1442534264.834865] lwj.4.info[2] lwj 4: node2: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello" ]
lt-flux-broker: [1442534264.835122] lwj.4.debug[2] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534264.835164] lwj.4.info[3] lwj 4: node3: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello" ]
lt-flux-broker: [1442534264.835362] lwj.4.debug[3] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534264.837158] lwj.4.debug[1] task0: pid 113264 (./mpi_hello): started
lt-flux-broker: [1442534264.838367] lwj.4.debug[2] task0: pid 77211 (./mpi_hello): started
lt-flux-broker: [1442534264.838536] lwj.4.debug[3] task0: pid 121727 (./mpi_hello): started
lt-flux-broker: [1442534264.840258] lwj.4.debug[0] updating job state to running
lt-flux-broker: [1442534264.953990] lwj.4.debug[0] task0: pid 242448 (./mpi_hello) exited with status 0x0000
0: completed MPI_Init in 0.007s.  There are 1 tasks
PMGR_COLLECTIVE ERROR: rank 1 on cab22:
lt-flux-broker: [1442534264.971842] lwj.4.debug[1] task0: pid 113264 (./mpi_hello) exited with status 0x008b
PMGR_COLLECTIVE ERROR: rank 3 on cab32:
lt-flux-broker: [1442534264.979419] lwj.4.debug[3] task0: pid 121727 (./mpi_hello) exited with status 0x008b
PMGR_COLLECTIVE ERROR: rank 2 on cab31:
lt-flux-broker: [1442534264.981217] lwj.4.debug[2] task0: pid 77211 (./mpi_hello) exited with status 0x008b
lt-flux-broker: [1442534264.982641] lwj.4.debug[0] updating job state to complete
lt-flux-broker: [1442534264.982923] lwj.4.info[1] job complete. exiting...
lt-flux-broker: [1442534264.982984] lwj.4.info[2] job complete. exiting...
lt-flux-broker: [1442534264.982681] lwj.4.info[3] job complete. exiting...
lt-flux-broker: [1442534264.984627] lwj.4.info[0] job complete. exiting...
wreckrun: tasks [1-3]: exited with signal 11
0: completed first barrier
0: completed MPI_Finalize
lt-flux-broker: [1442534264.991765] job.info[0] got request job.disconnect
```

Odd that one should work and the others die, I would have thought that even if the one got the right PMI it would have been given the wrong number of ranks for COMM_WORLD and failed, but it didn't.  Explicitly setting the pmi library for preload on all nodes results in a different error:

```
flux wreckrun -N 4 env LD_PRELOAD=../flux-core/src/lib/libpmi/.libs/libpmi.so ./mpi_hello
lt-flux-broker: [1442534350.140893] job.info[0] got request job.create
lt-flux-broker: [1442534350.142364] job.info[0] Setting job 5 to reserved
lt-flux-broker: waitpid: No child processes
lt-flux-broker: waitpid: No child processes
lt-flux-broker: [1442534350.182075] lwj.5.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442534350.183149] lwj.5.debug[0] lwj.5: node0: basis=0
lt-flux-broker: [1442534350.184621] lwj.5.info[0] lwj 5: node0: nprocs=1, nnodes=4, cmdline=[ "env", "LD_PRELOAD=..\/flux-core\/src\/lib\/libpmi\/.libs\/libpmi.so", ".\/mpi_hello" ]
lt-flux-broker: [1442534350.184685] lwj.5.debug[0] updating job state to starting
lt-flux-broker: [1442534350.185996] lwj.5.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534350.190197] lwj.5.debug[0] task0: pid 242481 (env): started
lt-flux-broker: [1442534350.229234] lwj.5.debug[1] initializing from CMB: rank=1
lt-flux-broker: [1442534350.230633] lwj.5.debug[1] lwj.5: node1: basis=1
lt-flux-broker: [1442534350.231312] lwj.5.info[1] lwj 5: node1: nprocs=1, nnodes=4, cmdline=[ "env", "LD_PRELOAD=..\/flux-core\/src\/lib\/libpmi\/.libs\/libpmi.so", ".\/mpi_hello" ]
lt-flux-broker: [1442534350.231510] lwj.5.debug[1] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534350.241098] lwj.5.debug[1] task0: pid 113284 (env): started
lt-flux-broker: [1442534350.281341] lwj.5.debug[3] initializing from CMB: rank=3
lt-flux-broker: [1442534350.283600] lwj.5.debug[3] lwj.5: node3: basis=3
lt-flux-broker: [1442534350.284630] lwj.5.info[3] lwj 5: node3: nprocs=1, nnodes=4, cmdline=[ "env", "LD_PRELOAD=..\/flux-core\/src\/lib\/libpmi\/.libs\/libpmi.so", ".\/mpi_hello" ]
lt-flux-broker: [1442534350.284854] lwj.5.debug[3] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534350.292973] lwj.5.debug[2] initializing from CMB: rank=2
lt-flux-broker: [1442534350.295110] lwj.5.debug[2] lwj.5: node2: basis=2
lt-flux-broker: [1442534350.294910] lwj.5.debug[3] task0: pid 121746 (env): started
lt-flux-broker: [1442534350.296157] lwj.5.info[2] lwj 5: node2: nprocs=1, nnodes=4, cmdline=[ "env", "LD_PRELOAD=..\/flux-core\/src\/lib\/libpmi\/.libs\/libpmi.so", ".\/mpi_hello" ]
lt-flux-broker: [1442534350.296435] lwj.5.debug[2] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442534350.306879] lwj.5.debug[2] task0: pid 77230 (env): started
lt-flux-broker: [1442534350.308262] lwj.5.debug[0] updating job state to running
lt-flux-broker: [1442534350.319638] lwj.5.debug[0] task0: pid 242481 (env) exited with status 0x0000
0: completed MPI_Init in 0.007s.  There are 1 tasks
mpi_hello: psm_entry.c:133: psm_doinit: Assertion `pg_rank < pg_size' failed.
lt-flux-broker: [1442534350.368861] lwj.5.debug[1] task0: pid 113284 (env) exited with status 0x0086
mpi_hello: psm_entry.c:133: psm_doinit: Assertion `pg_rank < pg_size' failed.
lt-flux-broker: [1442534350.425514] lwj.5.debug[3] task0: pid 121746 (env) exited with status 0x0086
mpi_hello: psm_entry.c:133: psm_doinit: Assertion `pg_rank < pg_size' failed.
lt-flux-broker: [1442534350.435959] lwj.5.debug[2] task0: pid 77230 (env) exited with status 0x0086
lt-flux-broker: [1442534350.437805] lwj.5.debug[0] updating job state to complete
lt-flux-broker: [1442534350.437984] lwj.5.info[2] job complete. exiting...
lt-flux-broker: [1442534350.438093] lwj.5.info[1] job complete. exiting...
lt-flux-broker: [1442534350.437657] lwj.5.info[3] job complete. exiting...
lt-flux-broker: [1442534350.439130] lwj.5.info[0] job complete. exiting...
wreckrun: tasks [1-3]: exited with signal 6
0: completed first barrier
0: completed MPI_Finalize
lt-flux-broker: [1442534350.447708] job.info[0] got request job.disconnect
```

This mpi_hello is built with the dotkit mvapich2_gnu.

Using the mvapich2-gnu-psm from opt however, hereafter mpi_hello_opt, is a bit different in that both versions of the multi-node test fail the same way the second does above.  So it gets the right PMI on all nodes, then dies anyway.  It did however print a bit more of a real error message:

```
mpi_hello_opt: psm_entry.c:133: psm_doinit: Assertion `pg_rank < pg_size' failed.
lt-flux-broker: [1442534840.676878] lwj.2.debug[2] task0: pid 77907 (env) exited with status 0x0086
Fatal error in MPI_Init: Internal MPI error!, error stack:
lt-flux-broker: [1442534840.696838] lwj.2.debug[1] task0: pid 113961 (env) exited with status 0x0100
MPIR_Init_thread(404):
MPID_Init(298).......: channel initialization failed
(unknown)(): Internal MPI error!
```

**Comment by garlick** [comment]
You might try setting `PMI_TRACE=255` in your environment to see that our PMI is being called.

Also check wreckrun `printenv FLUX_TASK_RANK FLUX_JOB_SIZE`.  wrexecd should be setting those for our PMI to pick up.

**Comment by grondo** [comment]
This is probably related to specifying only `-N4` on the command line. Can you try being explicit with `-n4 -N4`? Likely ntasks is using a wrong default and this is being propagated to PMI environment.

**Comment by dongahn** [comment]
I agree with Jim. I can think of various ways our pmi is not being called. DT_NEEDED of the binary could required a versioned pmi SO for example. libpmi.so.1

You may want to look at the actual SO dependency using an utility like ldd.

Then, make sure flux provides exactly that named pmi SO. You might also want to change the rpath of the mpi_hello using chrpath to point it to our directory.

**Comment by grondo** [comment]
This bug was probably introduced by decff66f, where `wreck.ntasks` is set to `nnodes` when `ntasks < nnodes` but the kvs is not updated. This stuff is all very ad-hoc right now, so a bit brittle. I've verified things work if I provide `-N4 -n4` for now.

BTW, for the `/opt` version of mvapich2-gnu-psm you should not need to use `LD_PRELOAD`. `LD_LIBRARY_PATH` is set appropriately (in a builddir at least), so an `/opt` version of mvapich2 should pick up flux libpmi because /opt compilers don't try to "help" by adding -rpath.

```
 grondo@hype248:~/tests/mpi$ which mpicc
/opt/mvapich2-gnu-psm-1.7/bin/mpicc
 grondo@hype248:~/tests/mpi$ mpicc -o mpihello mpihello.c 
 grondo@hype248:~/tests/mpi$  srun --pty ~/git/flux-core.git/src/cmd/flux broker  
[1442545680.327713] broker.info[0]: nodeset: [0-3] (complete)
[1442545680.327764] broker.info[0]: starting initial program
 grondo@hype248:~/tests/mpi$ flux setattr log-level 5
 grondo@hype248:~/tests/mpi$ flux wreckrun -N4 ./mpihello
rank 002 I am a worker: hype250 (rank=2/3)
rank 003 I am a worker: hype251 (rank=3/3)
rank 000 I am the master: hype248
rank 001 I am a worker: hype249 (rank=1/3)
 grondo@hype248:~/tests/mpi$ ldd ./mpihello | grep pmi
	libpmi.so.0 => /g/g0/grondo/git/flux-core.git/src/lib/libpmi/.libs/libpmi.so.0 (0x00002aaaab718000)
 grondo@hype248:~/tests/mpi$ exit
exit
 grondo@hype248:~/tests/mpi$ ldd mpihello | grep pmi
	libpmi.so.0 => /usr/lib64/libpmi.so.0 (0x00002aaaab718000)

```

**Comment by grondo** [comment]
I got 16 nodes on hype and verified up to 1024 tasks works fine when you get flux's PMI:
```
$ flux wreckrun -n 1024 ./mpi_hello
0: completed MPI_Init in 5.422s.  There are 1024 tasks
0: completed first barrier
0: completed MPI_Finalize
```

**Comment by grondo** [comment]
> though 3 seconds seems... unfortuante... for just four processes:

I can't reproduce this one, though I occasionally do see slowness. 
But most of the time:
```
 grondo@hype281:~/git/flux-core.git/src/test$ flux wreckrun -N4 -n4  ./mpi_hello 
0: completed MPI_Init in 0.036s.  There are 4 tasks
0: completed first barrier
0: completed MPI_Finalize

 grondo@hype281:~/git/flux-core.git/src/test$ flux wreckrun -N32 -n128  ./mpi_hello
0: completed MPI_Init in 0.737s.  There are 128 tasks
0: completed first barrier
0: completed MPI_Finalize

```

**Comment by trws** [comment]
Thanks for all the feedback on this.  I should have mentioned that if I use the mvapich from opt it normally starts quickly, but the one from dotkit pretty much always takes at least 2.75 seconds, which seems to be a result of using pmgr.  Somehow this makes me think PMGR is not a good default for our mpich builds...

Making sure that I always use `-n 4 -N 4` seems to get the job done, but losing the `-n 4` causes all kinds of problems, up to and including segfaulting or hanging a broker.  Should we have a separate issue for that?  Either way, using the `/opt` versions of MPI seems like by far the way to go right now, at least until we have another solution.

With everything working, the timing is a bit better, but including the finalize, it takes between .25 and  .5 seconds to execute 4 process hello on one node, and about 1.5 to execute it on four nodes, most of that in finalize for some reason.

Anyway, the title for this issue certainly is inaccurate.  Would you all prefer this be closed and open a separate issue for the tasks vs. nodes issue, or just re-title and keep this one around?

```
scogland at cab2 in ~/projects/flux/capacitor (master●)
$ time flux wreckrun -n 4 -N 4 ./mpi_hello_opt
lt-flux-broker: [1442595274.005888] job.info[0] got request job.create
lt-flux-broker: [1442595274.006665] job.info[0] Setting job 8 to reserved
lt-flux-broker: [1442595274.040939] lwj.8.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442595274.041576] lwj.8.debug[0] lwj.8: node0: basis=0
lt-flux-broker: [1442595274.042246] lwj.8.info[0] lwj 8: node0: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442595274.042310] lwj.8.debug[0] updating job state to starting
lt-flux-broker: [1442595274.043356] lwj.8.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442595274.047197] lwj.8.debug[0] task0: pid 150869 (./mpi_hello_opt): started
lt-flux-broker: [1442595274.145389] lwj.8.debug[1] initializing from CMB: rank=1
lt-flux-broker: [1442595274.146846] lwj.8.debug[1] lwj.8: node1: basis=1
lt-flux-broker: [1442595274.148222] lwj.8.debug[2] initializing from CMB: rank=2
lt-flux-broker: [1442595274.148296] lwj.8.info[1] lwj 8: node1: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442595274.148535] lwj.8.debug[1] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442595274.149028] lwj.8.debug[2] lwj.8: node2: basis=2
lt-flux-broker: [1442595274.149689] lwj.8.info[2] lwj 8: node2: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442595274.149917] lwj.8.debug[2] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442595274.157723] lwj.8.debug[1] task0: pid 134735 (./mpi_hello_opt): started
lt-flux-broker: [1442595274.158254] lwj.8.debug[2] task0: pid 103646 (./mpi_hello_opt): started
lt-flux-broker: [1442595274.160414] lwj.8.debug[3] initializing from CMB: rank=3
lt-flux-broker: [1442595274.162273] lwj.8.debug[3] lwj.8: node3: basis=3
lt-flux-broker: [1442595274.163666] lwj.8.info[3] lwj 8: node3: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442595274.163880] lwj.8.debug[3] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442595274.172293] lwj.8.debug[3] task0: pid 7101 (./mpi_hello_opt): started
lt-flux-broker: [1442595274.174266] lwj.8.debug[0] updating job state to running
0: completed MPI_Init in 0.174s.  There are 4 tasks
0: completed first barrier
0: completed MPI_Finalize
lt-flux-broker: [1442595275.343642] lwj.8.debug[0] task0: pid 150869 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595275.343531] lwj.8.debug[2] task0: pid 103646 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595275.344089] lwj.8.debug[1] task0: pid 134735 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595275.344056] lwj.8.debug[3] task0: pid 7101 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595275.346322] lwj.8.debug[0] updating job state to complete
lt-flux-broker: [1442595275.346667] lwj.8.info[2] job complete. exiting...
lt-flux-broker: [1442595275.346805] lwj.8.info[1] job complete. exiting...
lt-flux-broker: [1442595275.346749] lwj.8.info[3] job complete. exiting...
lt-flux-broker: [1442595275.348014] lwj.8.info[0] job complete. exiting...
flux wreckrun -n 4 -N 4 ./mpi_hello_opt  0.02s user 0.02s system 2% cpu 1.457 total
lt-flux-broker: [1442595275.350114] job.info[0] got request job.disconnect

scogland at cab2 in ~/projects/flux/capacitor (master●)
$ time flux wreckrun -n 4 ./mpi_hello_opt
lt-flux-broker: [1442595281.387249] job.info[0] got request job.create
lt-flux-broker: [1442595281.387767] job.info[0] Setting job 9 to reserved
lt-flux-broker: [1442595281.426081] lwj.9.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442595281.426613] lwj.9.debug[0] lwj.9: node0: basis=0
lt-flux-broker: [1442595281.428216] lwj.9.info[0] lwj 9: node0: nprocs=4, nnodes=1, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442595281.428275] lwj.9.debug[0] updating job state to starting
lt-flux-broker: [1442595281.429343] lwj.9.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442595281.432370] lwj.9.debug[0] task0: pid 150921 (./mpi_hello_opt): started
lt-flux-broker: [1442595281.432599] lwj.9.debug[0] task1: pid 150922 (./mpi_hello_opt): started
lt-flux-broker: [1442595281.432797] lwj.9.debug[0] task2: pid 150923 (./mpi_hello_opt): started
lt-flux-broker: [1442595281.433005] lwj.9.debug[0] task3: pid 150924 (./mpi_hello_opt): started
lt-flux-broker: [1442595281.434521] lwj.9.debug[0] updating job state to running
0: completed MPI_Init in 0.044s.  There are 4 tasks
0: completed first barrier
lt-flux-broker: [1442595281.591767] lwj.9.debug[0] task0: pid 150921 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595281.591992] lwj.9.debug[0] task2: pid 150923 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595281.592213] lwj.9.debug[0] task3: pid 150924 (./mpi_hello_opt) exited with status 0x0000
0: completed MPI_Finalize
lt-flux-broker: [1442595281.593582] lwj.9.debug[0] task1: pid 150922 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442595281.594449] lwj.9.debug[0] updating job state to complete
lt-flux-broker: [1442595281.596350] lwj.9.info[0] job complete. exiting...
flux wreckrun -n 4 ./mpi_hello_opt  0.01s user 0.02s system 11% cpu 0.253 total
lt-flux-broker: [1442595281.598665] job.info[0] got request job.disconnect
```

**Comment by grondo** [comment]
Let's retitle and keep this issue. I've got a fix for `flux-wreckrun` (I think) but need to figure out how to write some tests.

On PMGR, I'm not even sure how that version is working with flux. Is it connecting back to srun somehow?  We should probably figure out if that is the case and determine how to prevent it. Running a non PMI PMI under flux wreckrun should cause a hard or other easily detectable failure.

For timings, you might want to run with `-v -o stdio-delay-commit`. The actual job is probably completing long before all the I/O is done and pushed to the kvs. The extra `-v` will display timing output in wreckrun for when the job actually goes from starting->running->complete.

**Comment by trws** [comment]
Ok, I'll retitle momentarily.  As to the timing, that's a good point, and I just retried it.  Full output is below, but the relevant parts are, I think and please correct me if I'm mistaken, the task 0 starting at 1442599314.135360 and 2 dying last at 1442599316.343882, which when reduced out from the initial timestamp are 0.36 and 2.56.  So, it takes about half a second to launch everything, the MPI processes take about 2 seconds to run on four nodes, and then it takes another half second or so to finalize everything (in this run, the numbers are larger).

By the way, PMGR might be finding something on the local node, since the task on the launching node works but each one on the other nodes fails.  I'm not sure what that is, but completely agree we should mask it out if we can.


```
scogland at cab2 in ~/projects/flux/capacitor (master●)
$ time flux wreckrun -v -o stdio-delay-commit -n 4 -N 4 ./mpi_hello_opt
wreckrun: 0.002s: Sending LWJ request for 4 tasks (cmdline "./mpi_hello_opt")
lt-flux-broker: [1442599313.777741] job.info[0] got request job.create
lt-flux-broker: [1442599313.786213] job.info[0] Setting job 2448 to reserved
wreckrun: 0.033s: Registered jobid 2448
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.048s: Sending run event
wreckrun: 0.168s: State = reserved
lt-flux-broker: [1442599314.058030] lwj.2448.debug[2] initializing from CMB: rank=2
lt-flux-broker: [1442599314.058299] lwj.2448.debug[2] Setting option stdio-delay-commit = 1
lt-flux-broker: [1442599314.058829] lwj.2448.debug[2] lwj.2448: node2: basis=2
lt-flux-broker: [1442599314.059429] lwj.2448.info[2] lwj 2448: node2: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442599314.059697] lwj.2448.debug[2] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442599314.070604] lwj.2448.debug[2] task0: pid 151758 (./mpi_hello_opt): started
lt-flux-broker: [1442599314.099494] lwj.2448.debug[1] initializing from CMB: rank=1
lt-flux-broker: [1442599314.100462] lwj.2448.debug[3] initializing from CMB: rank=3
lt-flux-broker: [1442599314.100444] lwj.2448.debug[1] Setting option stdio-delay-commit = 1
lt-flux-broker: [1442599314.101361] lwj.2448.debug[3] Setting option stdio-delay-commit = 1
lt-flux-broker: [1442599314.100999] lwj.2448.debug[1] lwj.2448: node1: basis=1
lt-flux-broker: [1442599314.101580] lwj.2448.info[1] lwj 2448: node1: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442599314.102112] lwj.2448.debug[3] lwj.2448: node3: basis=3
lt-flux-broker: [1442599314.101816] lwj.2448.debug[1] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442599314.102812] lwj.2448.info[3] lwj 2448: node3: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442599314.103058] lwj.2448.debug[3] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
lt-flux-broker: [1442599314.108027] lwj.2448.debug[0] initializing from CMB: rank=0
lt-flux-broker: [1442599314.108464] lwj.2448.debug[0] Setting option stdio-delay-commit = 1
lt-flux-broker: [1442599314.109038] lwj.2448.debug[0] lwj.2448: node0: basis=0
lt-flux-broker: [1442599314.110025] lwj.2448.info[0] lwj 2448: node0: nprocs=1, nnodes=4, cmdline=[ ".\/mpi_hello_opt" ]
lt-flux-broker: [1442599314.110090] lwj.2448.debug[0] updating job state to starting
lt-flux-broker: [1442599314.120318] lwj.2448.debug[0] reading lua files from /g/g12/scogland/projects/flux/flux-core/src/modules/wreck/lua.d/*.lua
wreckrun: 0.346s: State = starting
lt-flux-broker: [1442599314.135360] lwj.2448.debug[0] task0: pid 230772 (./mpi_hello_opt): started
lt-flux-broker: [1442599314.144652] lwj.2448.debug[1] task0: pid 194110 (./mpi_hello_opt): started
lt-flux-broker: [1442599314.160413] lwj.2448.debug[3] task0: pid 82903 (./mpi_hello_opt): started
lt-flux-broker: [1442599314.168718] lwj.2448.debug[0] updating job state to running
wreckrun: 0.407s: State = running
lt-flux-broker: [1442599315.342329] lwj.2448.debug[0] task0: pid 230772 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442599315.342821] lwj.2448.debug[3] task0: pid 82903 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442599315.343132] lwj.2448.debug[1] task0: pid 194110 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442599316.343882] lwj.2448.debug[2] task0: pid 151758 (./mpi_hello_opt) exited with status 0x0000
lt-flux-broker: [1442599316.352996] lwj.2448.debug[0] updating job state to complete
0: completed first barrier
0: completed MPI_Finalize
lt-flux-broker: [1442599316.356273] lwj.2448.info[3] job complete. exiting...
lt-flux-broker: [1442599316.356003] lwj.2448.info[1] job complete. exiting...
lt-flux-broker: [1442599316.356404] lwj.2448.info[2] job complete. exiting...
0: completed MPI_Init in 0.085s.  There are 4 tasks
lt-flux-broker: [1442599316.366325] lwj.2448.info[0] job complete. exiting...
wreckrun: 2.594s: State = complete
wreckrun: All tasks completed successfully.
flux wreckrun -v -o stdio-delay-commit -n 4 -N 4 ./mpi_hello_opt  0.01s user 0.02s system 1% cpu 2.640 total
lt-flux-broker: [1442599316.372141] job.info[0] got request job.disconnect
```

**Comment by trws** [comment]
Since it's MPI, I took a quick tau trace (still working on something else, but perf works on cab, so that's something).  This is where the time is going:

```
-------------------------------------------------------------------------------
Thread: n,c,t 0,0,0
-------------------------------------------------------------------------------
 excl.secs  excl.%   cum.%     calls  function
     2.012   82.5%   82.5%         1  MPI_Finalize()
     0.419   17.2%   99.7%         1  MPI_Init()
     0.008    0.3%  100.0%         1  .TAU application
   1.3E-05    0.0%  100.0%         1  MPI_Barrier()
         0    0.0%  100.0%         1  MPI_Comm_rank()
         0    0.0%  100.0%         1  MPI_Comm_size()
```
Almost all in finalize.

Perf tells me that nearly all of that is in the PSM layer, so it's waiting on comms.  A merged perf with broker 0 and mpi_hello_opt rank 0 makes it look like the delay might actually be on the mvapich side, since flux is showing as a minor player in this.  I'm not sure why that would be though.

```
 44.51%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] psmi_cycles_left                                                                                                                                    ◆
 15.65%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] ips_ptl_poll                                                                                                                                        ▒
  2.87%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] ips_proto_fini                                                                                                                                      ▒
  2.69%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] __psmi_poll_internal                                                                                                                                ▒
  1.74%  flux-broker-0  libzmq.so.4.0.0                    [.] 0x0000000000017bbf                                                                                                                                  ▒
  1.29%  flux-broker-0  libczmq.so.1.1.0                   [.] 0x0000000000019b83                                                                                                                                  ▒
  1.13%  flux-broker-0  [kernel.kallsyms].exit.text        [k] 0x000000001f6bd939                                                                                                                                  ▒
  0.84%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] psmi_cycles_left@plt                                                                                                                                ▒
  0.58%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tc_calloc                                                                                                                                           ▒
  0.56%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] psmi_poll_internal@plt                                                                                                                              ▒
  0.54%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tc_free                                                                                                                                             ▒
  0.51%  flux-broker-0  [kernel.kallsyms]                  [k] _spin_lock                                                                                                                                          ▒
  0.50%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] __psmi_poll_noop                                                                                                                                    ▒
  0.43%  flux-broker-0  libc-2.12.so                       [.] memcpy                                                                                                                                              ▒
  0.40%  flux-broker-0  [kernel.kallsyms]                  [k] _spin_lock_irqsave                                                                                                                                  ▒
  0.39%  flux-broker-0  [kernel.kallsyms]                  [k] system_call                                                                                                                                         ▒
  0.39%  flux-broker-0  libjson-c.so.2.0.1                 [.] lh_char_hash                                                                                                                                        ▒
  0.35%  mpi_hello_opt  libpsm_infinipath.so.1.14          [.] ips_proto_disconnect                                                                                                                                ▒
  0.35%  flux-broker-0  libjson-c.so.2.0.1                 [.] json_escape_str                                                                                                                                     ▒
  0.34%  flux-broker-0  libc-2.12.so                       [.] vfprintf                                                                                                                                            ▒
  0.34%  flux-broker-0  [kernel.kallsyms]                  [k] fget_light                                                                                                                                          ▒
  0.33%  flux-broker-0  [kernel.kallsyms]                  [k] do_sys_poll                                                                                                                                         ▒
  0.33%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tcmalloc::CentralFreeList::FetchFromSpans()                                                                                                         ▒
  0.28%  flux-broker-0  libjson-c.so.2.0.1                 [.] json_tokener_parse_ex                                                                                                                               ▒
  0.27%  mpi_hello_opt  [kernel.kallsyms]                  [k] clear_page_c                                                                                                                                        ▒
  0.26%  flux-broker-0  [kernel.kallsyms]                  [k] copy_user_generic_string                                                                                                                            ▒
  0.25%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tc_malloc                                                                                                                                           ▒
  0.24%  flux-broker-0  [kernel.kallsyms]                  [k] _atomic_dec_and_lock                                                                                                                                ▒
  0.23%  flux-broker-0  [kernel.kallsyms]                  [k] update_curr                                                                                                                                         ▒
  0.23%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned long, int)                                                  ▒
  0.23%  mpi_hello_opt  [kernel.kallsyms]                  [k] _spin_lock                                                                                                                                          ▒
  0.20%  flux-broker-0  libtcmalloc_and_profiler.so.4.1.0  [.] tcmalloc::CentralFreeList::ReleaseToSpans(void*)                                                                                                    ▒
  0.19%  flux-broker-0  libc-2.12.so                       [.] __memset_sse2
```

**Comment by grondo** [comment]
Thanks for new title! Want to cut and paste the perf and tau outputs above into a new issue? Seems like relevant data to keep around, but unrelated to the current issues.

Also, I'm wondering if "PMGR occasionally pretends to work" should also be its own issue?
